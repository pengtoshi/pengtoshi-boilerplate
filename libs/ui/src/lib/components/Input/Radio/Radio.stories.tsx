import type { StoryFn } from "@storybook/react";
import { useState } from "react";
import type { ComponentProps } from "react";
import { Radio } from "./Radio";

export default { component: Radio };

const Template: StoryFn<ComponentProps<typeof Radio>> = (args: ComponentProps<typeof Radio>) => {
  const [active, setActive] = useState(false);
  return <Radio {...args} active={active} handleClick={() => setActive(!active)} />;
};

export const Default = Template.bind({});
Default.args = {
  size: "medium",
};
