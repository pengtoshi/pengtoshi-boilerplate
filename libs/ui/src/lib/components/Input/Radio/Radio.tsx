import clsx from "clsx";
import { useEffect, useState } from "react";
import type { UIProps } from "../../../props";
import { Interaction } from "../../Display/Interaction/Interaction";

export interface RadioProps extends UIProps.Button {
  active: boolean;
  handleClick: () => void;
  size?: "small" | "medium";
  disabled?: boolean;
}

const RadioSize = {
  small: "h-5 w-5",
  medium: "h-6 w-6",
};

const InteractionSize = {
  small: "h-7 w-7",
  medium: "h-8 w-8",
};

export const Radio = ({ active, handleClick, size = "medium", className, ...props }: RadioProps) => {
  const [internalActive, setInternalActive] = useState(active);

  useEffect(() => {
    setInternalActive(active);
  }, [active]);

  return (
    <button
      type="button"
      className={clsx("group relative p-0.5 disabled:opacity-40", RadioSize[size], className)}
      {...props}
      onClick={() => {
        setInternalActive(!internalActive);
        handleClick();
      }}
    >
      <div
        className={clsx(
          "h-full w-full rounded-full bg-gray-100",
          internalActive ? "border-4 border-primary-500" : "border-[1.5px] border-gray-300",
        )}
      />
      <Interaction focus={false} className={clsx("-left-1 -top-1 rounded-full", InteractionSize[size])} />
    </button>
  );
};
