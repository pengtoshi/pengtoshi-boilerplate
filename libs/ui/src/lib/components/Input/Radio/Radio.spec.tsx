import { composeStories } from "@storybook/react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import * as stories from "./Radio.stories";

describe("Radio", () => {
  Object.entries(composeStories(stories)).forEach(([, Story]) => {
    it("should render", async () => {
      const { baseElement } = render(<Story />);
      expect(baseElement).toMatchSnapshot();
    });
  });
});
