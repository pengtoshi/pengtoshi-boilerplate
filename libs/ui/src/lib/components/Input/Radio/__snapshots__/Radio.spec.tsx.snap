// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should render 1`] = `
<body>
  <div>
    <section
      aria-atomic="false"
      aria-label="Notifications alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      tabindex="-1"
    />
    <button
      class="group relative p-0.5 disabled:opacity-40 h-6 w-6"
      type="button"
    >
      <div
        class="h-full w-full rounded-full bg-gray-50 border-[1.5px] border-gray-300"
      />
      <div
        class="absolute inset-0 bg-gray-950 opacity-0 transition-opacity duration-300 group-hover:opacity-[0.06] group-active:opacity-[0.12] -left-1 -top-1 rounded-full h-8 w-8"
      />
    </button>
  </div>
</body>
`;
